### ISSUE TITLE
Should return copy of internal list in Names object

### ISSUE THREAD

**Issue by DiDeoxy** [BiocPy/BiocUtils#18]
https://github.com/BiocPy/BiocUtils/blob/9a64a65172d8771bab5539af3edb26bea902cbdf/src/biocutils/Names.py#L97-L102

By returning the internal list any modifications to this list also occur in the Names object. This can cause unintended side effects if, for example, the Names object is used in a parent object like a BiocFrame. Should probably return a copy instead?

**Comment by LTLA** [comment]
I think the behavior is correct under **BiocPy**'s current functional-style policy. Generally, mutations are not expected to happen outside of the various `.set_*()` functions (with some exceptions for back-compatible setters), so the fact that `.as_list()` returns a reference to an internal member is not a problem, as it shouldn't be getting modified anyway.

That said, users will probably not respect the read-only expectation on the return values, so it wouldn't hurt to have a `copy = true` default that can be disabled for performance. This mirrors how the **scran.js** classes enforce read-only access to Wasm memory from Javascript.

In the meantime, I would advise just making a copy.

**Comment by LTLA** [comment]
After some consideration, I remembered why we didn't provide a `copy=` argument in the first place.

The first issue is that of consistency. If we set `copy=True` as the safe default here, we should do so for all **BiocPy** getters. This may not be cheap, e.g., for the `assay` getters in a `SummarizedExperiment`.

The second issue is the definition of a "copy". To achieve the desired safety, any copy must be deep, as a user could modify any aspect of the returned object. This further adds to the expenses of a hypothetical `copy=True` default.

Furthermore, if we really wanted to be safe, we would also need to make (deep) copies of any data passed via the class constructors. This is because the caller would likely still be holding a reference to the same objects, and any mutations via the caller's reference would result in unintended changes to the class's internal data.

Note, none of this is a problem when returning **BiocPy** objects, as the functional policy discourages in-place modifications. But we can't do much about Python's basic types, and we're unlikely to ever get `const`-ness concepts in the language, so we need to choose between defaulting to safety or efficiency.

Probably the best solution here is to just document all getters as read-only and instruct users to make a (deep) copy if they intend to modify the return value. This is possibly also the most Pythonic solution given the language's "consenting adults" philosophy.
